### ISSUE TITLE
sha512_acc START_ADDRESS behavior does not match documentation

### ISSUE THREAD

**Issue by korran** [chipsalliance/caliptra-rtl#55]
From https://chipsalliance.github.io/caliptra-rtl/main/external-regs/?p=caliptra_top_reg.sha512_acc_csr.START_ADDRESS:

> The start address for FW controlled SHA performed on data stored in the mailbox. Start Address must be dword aligned.

This implies that the address is a byte address, not a word address. However, the RTL seems to be treating this register as though it were a word address:

![sha512acc set_data_len](https://user-images.githubusercontent.com/993854/230236164-502f3116-7331-4559-9251-fe5c6947c700.png)

![sha512acc read_mbox](https://user-images.githubusercontent.com/993854/230236199-31807254-47ed-46b6-bd5b-6e752629d2c6.png)

Because of this, the boot ROM fails to validate the firmware image when running inside verilator.

Either the RTL needs to be changed to match the documentation, or the documentation (and sw-emulator) need to be changed to match the RTL. Please advise.

**Comment by bharatpillilli** [comment]
Adding Mike

**Comment by Nitsirks** [comment]
It sounds like the easiest fix is to keep the spec and sw emulator and fix RTL so that it converts byte address to mailbox read pointer.

I'll submit that fix today.

**Comment by korran** [comment]
Thanks! Can you also include it in a branch somewhere in the public repo so I can use it to unblock my boot-ROM debugging?

**Comment by Nitsirks** [comment]
Try this branch please: michnorris-msft-issue55

**Comment by korran** [comment]
Looks great! Thanks so much!

```console
$ git log -p -1
commit 04cd005339f7bb77927fe48bc8ff10046c501642 (HEAD -> rom-debug)
Author: Kor Nielsen <kor@google.com>
Date:   Wed Apr 5 16:44:09 2023 -0700

    Add test case for #55.

diff --git a/drivers/test-fw/src/bin/sha384acc_tests.rs b/drivers/test-fw/src/bin/sha384acc_tests.rs
index 1408747..e7146bd 100644
--- a/drivers/test-fw/src/bin/sha384acc_tests.rs
+++ b/drivers/test-fw/src/bin/sha384acc_tests.rs
@@ -105,6 +105,35 @@ fn test_digest2() {
     }
 }
 
+fn test_digest_offset() {
+    let data = "abcdefghijklmnopqrst".as_bytes();
+    let expected: [u8; SHA384_HASH_SIZE] = [
+        0xd4, 0xcc, 0x9a, 0x0d, 0xc5, 0x46, 0x09, 0x40, 0xb0, 0x50, 0xa2,
+        0x42, 0x14, 0xf6, 0x78, 0xf6, 0x3b, 0x99, 0x3e, 0xc3, 0xc5, 0x7d, 0xb9,
+        0xcc, 0x20, 0x7b, 0x20, 0x9c, 0xbd, 0xa7, 0xcc, 0x09, 0xe9, 0x4a, 0x84,
+        0x62, 0x83, 0x56, 0x7d, 0x28, 0xd8, 0xc7, 0x73, 0xc1, 0x87, 0x39, 0x07,
+        0xa7,
+    ];
+
+    let mut digest = Array4x12::default();
+    let sha_acc = Sha384Acc::default();
+
+    if let Some(mut txn) = Mailbox::default().try_start_send_txn() {
+        const CMD: u32 = 0x1c;
+        assert!(txn.send_request(CMD, &data).is_ok());
+
+        if let Some(mut sha_acc_op) = sha_acc.try_start_operation() {
+            let result = sha_acc_op.digest(8, 4, false, (&mut digest).into());
+            assert!(result.is_ok());
+            assert_eq!(digest, Array4x12::from(expected));
+            drop(sha_acc_op);
+        } else {
+            assert!(false);
+        }
+        drop(txn);
+    }
+}
+
 fn test_digest_zero_size_buffer() {
     let expected: [u8; SHA384_HASH_SIZE] = [
         0x38, 0xB0, 0x60, 0xA7, 0x51, 0xAC, 0x96, 0x38, 0x4C, 0xD9, 0x32, 0x7E, 0xB1, 0xB1, 0xE3,
$ cargo test -p caliptra-drivers test_sha384acc --features=verilator  -- --nocapture
UART: sha384acc::test_digest_offset...	[failed]
UART: Error: panicked at 'assertion failed: `(left == right)`
UART: Error: panicked at 'assertion failed: `(left == right)`
UART:   left: `Array4xN([2877944584, 4120666457, 2145879868, 3485676380, 3614955051, 1694951571, 3821374125, 516462036, 4102583958, 1531478451, 527554566, 4100025493])`,
UART:  right: `Array4xN([3570178573, 3309701440, 2958074434, 351697142, 999898819, 3313351116, 544940188, 3181890569, 3913974882, 2203483432, 3636949953, 2268661671])`', drivers/test-fw/src/bin/sha384acc_tests.rs:128:13
UART: 
* TESTCASE FAILED
$ (cd hw-latest/caliptra-rtl && git fetch && git checkout origin/michnorris-msft-issue55)
remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Total 6 (delta 5), reused 6 (delta 5), pack-reused 0
Unpacking objects: 100% (6/6), 607 bytes | 46.00 KiB/s, done.
From https://github.com/chipsalliance/caliptra-rtl
 * [new branch]      michnorris-msft-issue55 -> origin/michnorris-msft-issue55
Previous HEAD position was cdd9afa Fix AHB multiplexing issue: #20
HEAD is now at 18741f5 Aligning start address from byte address to mailbox pointer to align with register and spec description Fixes github issue 55
$ cargo test -p caliptra-drivers test_sha384acc --features=verilator  -- --nocapture
UART: sha384acc::test_digest_offset...	[ok]
* TESTCASE PASSED
test test_sha384acc ... ok

```

**Comment by korran** [comment]
Ok, fix is merged into caliptra-sw's rtl submodule as of chipsalliance/caliptra-sw#147

**Comment by howardtr** [comment]
https://github.com/chipsalliance/caliptra-rtl/commit/28e8772f7dc9b96850df14ae83c6c4804ac48e06 has made it to main
